internal/mempool/v1/mempool.go

```diff
@@ -450,422 +416,356 @@ func (txmp *TxMempool) Update(
-		txmp.logger.Error("failed to flush transactions during rechecking", "err", err)
+		txmp.logger.Error("failed to flush transactions during recheck", "err", err)
 	}
 }
 
 // canAddTx returns an error if we cannot insert the provided *WrappedTx into
-// the mempool due to mempool configured constraints. Otherwise, nil is returned
-// and the transaction can be inserted into the mempool.
+// the mempool due to mempool configured constraints. Otherwise, nil is
+// returned and the transaction can be inserted into the mempool.
 func (txmp *TxMempool) canAddTx(wtx *WrappedTx) error {
-	var (
-		numTxs    = txmp.Size()
-		sizeBytes = txmp.SizeBytes()
-	)
+	numTxs := txmp.Size()
+	txBytes := txmp.SizeBytes()
 
-	if numTxs >= txmp.config.Size || int64(wtx.Size())+sizeBytes > txmp.config.MaxTxsBytes {
+	if numTxs >= txmp.config.Size || wtx.Size()+txBytes > txmp.config.MaxTxsBytes {
 		return types.ErrMempoolIsFull{
 			NumTxs:      numTxs,
 			MaxTxs:      txmp.config.Size,
-			TxsBytes:    sizeBytes,
+			TxsBytes:    txBytes,
 			MaxTxsBytes: txmp.config.MaxTxsBytes,
 		}
 	}
 
 	return nil
 }
 
-func (txmp *TxMempool) insertTx(wtx *WrappedTx) {
-	txmp.txStore.SetTx(wtx)
-	txmp.priorityIndex.PushTx(wtx)
-	txmp.heightIndex.Insert(wtx)
-	txmp.timestampIndex.Insert(wtx)
-
-	// Insert the transaction into the gossip index and mark the reference to the
-	// linked-list element, which will be needed at a later point when the
-	// transaction is removed.
-	gossipEl := txmp.gossipIndex.PushBack(wtx)
-	wtx.gossipEl = gossipEl
-
-	atomic.AddInt64(&txmp.sizeBytes, int64(wtx.Size()))
-}
-
-func (txmp *TxMempool) removeTx(wtx *WrappedTx, removeFromCache bool) {
-	if txmp.txStore.IsTxRemoved(wtx.hash) {
-		return
-	}
-
-	txmp.txStore.RemoveTx(wtx)
-	txmp.priorityIndex.RemoveTx(wtx)
-	txmp.heightIndex.Remove(wtx)
-	txmp.timestampIndex.Remove(wtx)
-
-	// Remove the transaction from the gossip index and cleanup the linked-list
-	// element so it can be garbage collected.
-	txmp.gossipIndex.Remove(wtx.gossipEl)
-	wtx.gossipEl.DetachPrev()
-
-	atomic.AddInt64(&txmp.sizeBytes, int64(-wtx.Size()))
-
-	if removeFromCache {
-		txmp.cache.Remove(wtx.tx)
-	}
-}
-
-// purgeExpiredTxs removes all transactions that have exceeded their respective
-// height and/or time based TTLs from their respective indexes. Every expired
-// transaction will be removed from the mempool entirely, except for the cache.
+// purgeExpiredTxs removes all transactions from the mempool that have exceeded
+// their respective height or time-based limits as of the given blockHeight.
+// Transactions removed by this operation are not removed from the cache.
 //
-// NOTE: purgeExpiredTxs must only be called during TxMempool#Update in which
-// the caller has a write-lock on the mempool and so we can safely iterate over
-// the height and time based indexes.
+// The caller must hold txmp.mtx exclusively.
 func (txmp *TxMempool) purgeExpiredTxs(blockHeight int64) {
```

Comment:
Maybe a premature optimisation but if we add 
```
if txmp.config.TTLNumBlocks == 0 && txmp.config.TTLDuration {
    return
}
```
then we can skip looping through all the elements. I think the default is actually that these TTLs are switched off but maybe double check.

Reply:
That would make the code more legible anyway, so it's worth doing even though the savings may be negligible.